AmplifyPlugins/Auth/AWSAuthPluginTests/Utils/AuthUserAttributeKeyTests.swift

```diff
@@ -0,0 +1,73 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import XCTest
+import Amplify
+import AWSAuthPlugin
+
+class AuthUserAttributeKeyTests: XCTestCase {
+
+    /// Test if attribute raw value gives us the right value
+    ///
+    /// - Given: Any attribute from the enum AuthUserAttributeKey say .phoneNumber
+    /// - When:
+    ///    - I get rawValue of the enum
+    /// - Then:
+    ///    - The value should be equal to the one defined by Cognito.
+    ///    https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html
+    ///
+    func testSuccessfulRawValue() {
+        let attribute = AuthUserAttributeKey.phoneNumber
+        XCTAssertEqual(attribute.rawValue, "phone_number")
+    }
+
+    /// Test if initializing through a valid raw string gives back an enum
+    ///
+    /// - Given: A valid raw string from the Cognito
+    /// - When:
+    ///    - I create an enum using the raw value
+    /// - Then:
+    ///    - The enum should be mapping to the exact value
+    ///
+    func testInitWithRawValue() {
+        let attribute = AuthUserAttributeKey(rawValue: "middle_name")
+        XCTAssertEqual(attribute, .middleName)
+    }
+
+    /// Test if initializing through a custom raw string gives back an enum
+    ///
+    /// - Given: A custom raw string to be used as custom attribute
+    /// - When:
+    ///    - I create an enum using the raw value
+    /// - Then:
+    ///    - The enum should be custom attribute type
+    ///
+    func testInitWithCustomRawValue() {
+        let attribute = AuthUserAttributeKey(rawValue: "someattribute")
+        switch attribute {
+        case .custom(let attribute):
+            print(attribute)
+        default:
+            XCTFail("Attribute type should be custom")
+        }
+    }
+
+    /// Test if attribute raw value gives us the right value for a custom attribute
+    ///
+    /// - Given: A custom attribute from AuthUserAttributeKey
+    /// - When:
+    ///    - I get rawValue of the enum
+    /// - Then:
+    ///    - The value should be equal to "custom: <value given>"
+    ///
+    func testSuccessfulCustomRawValue() {
+        let customAttributeKey = "someattribute2"
+        let attribute = AuthUserAttributeKey.custom(customAttributeKey)
+        XCTAssertEqual(attribute.rawValue, "custom: \(customAttributeKey)")
```

Comment:
i'm not entirely sure when there is a custom attribute key, maybe when the service updates and library is not yet handling the new key, then it would get created as `.custom(key)`. then if it's forward compatible, i would expect the library to just get the attributeKey as is as a string `<key>`. Using `custom: <key>` will break since the service doesn't know what to do with it

Reply:
Explained above